src/System.CommandLine/Parsing/ParseResultVisitor.cs

```diff
@@ -153,88 +153,78 @@ protected override void Stop(SyntaxNode node)
 
-                foreach (var result in commandResult.Children)
+            foreach (var result in _innermostCommandResult.Children)
+            {
+                switch (result)
                 {
-                    switch (result)
-                    {
-                        case ArgumentResult argumentResult:
-
-                            ValidateArgument(argumentResult);
+                    case ArgumentResult argumentResult:
 
-                            break;
+                        ValidateArgumentResult(argumentResult);
 
-                        case OptionResult optionResult:
-
-                            var argument = optionResult.Option.Argument;
+                        break;
 
-                            var arityFailure = ArgumentArity.Validate(
-                                optionResult,
-                                argument,
-                                argument.Arity.MinimumNumberOfValues,
-                                argument.Arity.MaximumNumberOfValues);
+                    case OptionResult optionResult:
 
-                            if (arityFailure != null)
-                            {
-                                _errors.Add(
-                                    new ParseError(arityFailure.ErrorMessage, optionResult));
-                            }
+                        ValidateOptionResult(optionResult);
 
-                            var results = optionResult.Children
-                                                      .OfType<ArgumentResult>()
-                                                      .ToArray();
+                        break;
+                }
+            }
+        }
 
-                            foreach (var a in results)
-                            {
-                                ValidateArgument(a);
-                            }
+        private void ValidateCommandResult(CommandResult commandResult)
+        {
+            if (commandResult.Command is Command command)
+            {
+                foreach (var validator in command.Validators)
+                {
+                    var errorMessage = validator(commandResult);
 
-                            break;
+                    if (!string.IsNullOrWhiteSpace(errorMessage))
+                    {
+                        _errors.Add(
+                            new ParseError(errorMessage, commandResult));
                     }
                 }
             }
-        }
 
-        private void ValidateCommand(CommandResult commandResult)
-        {
-            foreach (var a in commandResult
-                              .Command
-                              .Arguments)
+            foreach (var option in commandResult
+                                   .Command
+                                   .Options)
             {
-                if (a is Argument argument)
+                if (option is Option o &&
+                    o.Required && 
```

Comment:
Should `Required` be added to `IOption` to avoid the need for the `Option` cast here? Thouhg looking at the implementation, it appears that the only children that will be returned in the `Command.Options` will be `Options`.

Reply:
This is still a little awkward but maybe, yes. `IOption` is, in practice, a sort of read-only `Option`, but we didn't go to the additional step of making it a truly immutable copy of the original mutable `Option`.